test/images/image-util.sh

```diff
@@ -117,23 +141,46 @@ push() {
   docker_version_check
   TAG=$(<"${IMAGE}"/VERSION)
   if [[ -f ${IMAGE}/BASEIMAGE ]]; then
-    archs=$(listArchs)
+    base_images=$(listBaseImages)
+    # NOTE(claudiub): if the REMOTE_DOCKER_URL var is not set, or it is an empty string, we must skip
+    # pushing the Windows image and including it into the manifest list.
+    if [[ ! -v "REMOTE_DOCKER_URL" || -z "${REMOTE_DOCKER_URL}" ]]; then
+      if [[ ! -z "$(printf "%s\n" $base_images | grep '^windows')" ]]; then
+        echo "Skipping pushing the image '${IMAGE}' for Windows. REMOTE_DOCKER_URL should be set, containing the URL to a Windows docker daemon."
+        base_images=$(printf "%s\n" $base_images | grep -v "^windows")
+      fi
+    fi
   else
-    archs=${!QEMUARCHS[@]}
+    # prepend linux/ to the QEMUARCHS items.
+    base_images=`printf 'linux/%s\n' "${!QEMUARCHS[@]}"`
   fi
-  for arch in ${archs}; do
-    docker push "${REGISTRY}/${IMAGE}-${arch}:${TAG}"
+  for base_image in ${base_images}; do
+    if [[ $base_image =~ .*/.* ]]; then
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
+
+    if [[ "$os_name" = "linux" ]]; then
+      docker push "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
+    else
+      # NOTE(claudiub): We're pushing the image we built on the remote Windows node.
+      docker --tlsverify -H "${REMOTE_DOCKER_URL}" push "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
+    fi
   done
 
   kube::util::ensure-gnu-sed
 
   # The manifest command is still experimental as of Docker 18.09.2
   export DOCKER_CLI_EXPERIMENTAL="enabled"
-  # Make archs list into image manifest. Eg: 'amd64 ppc64le' to '${REGISTRY}/${IMAGE}-amd64:${TAG} ${REGISTRY}/${IMAGE}-ppc64le:${TAG}'
-  manifest=$(echo "$archs" | ${SED} -e "s~[^ ]*~$REGISTRY\/$IMAGE\-&:$TAG~g")
+  # Make base_images list into image manifest. Eg: 'linux/amd64 linux/ppc64le' to '${REGISTRY}/${IMAGE}:${TAG}-linux-amd64 ${REGISTRY}/${IMAGE}:${TAG}-linux-ppc64le'
+  manifest=$(echo "$base_images" | ${SED} "s~\/~-~g" | ${SED} -e "s~[^ ]*~$REGISTRY\/$IMAGE:$TAG\-&~g")
   docker manifest create --amend "${REGISTRY}/${IMAGE}:${TAG}" ${manifest}
-  for arch in ${archs}; do
-    docker manifest annotate --arch "${arch}" "${REGISTRY}/${IMAGE}:${TAG}" "${REGISTRY}/${IMAGE}-${arch}:${TAG}"
+  for base_image in ${base_images}; do
+    if [[ $base_image =~ .*/.* ]]; then
+      os_name=`echo $base_image | cut -d "/" -f 1`
+      arch=`echo $base_image | cut -d "/" -f 2`
+    fi
+    docker manifest annotate --os "${os_name}" --arch "${arch}" "${REGISTRY}/${IMAGE}:${TAG}" "${REGISTRY}/${IMAGE}:${TAG}-${os_name}-${arch}"
```

Comment:
> The previous naming template would generate a plethora of images (Ai * N images, where Ai is the number of OS/architectures for the image i and N is the number of images), while the new naming template will reduce the number of images to N.

Could you clarify the intent a bit on this? In the end you still have `#os X #arch X #tag` number of images/tags. Before you'd have `img-windows/amd64:tag` and now you have `img:tag-windows-amd64`. In either case docker will be reutilzing the layers so isn't the end result the same?